File: sdk/go/transport.go
package aghsdk

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
	"sync"
)

const (
	transportErrorKey = "error"
)

// DefaultMaxMessageBytes is the default JSON-RPC line size limit.
const DefaultMaxMessageBytes = 10 * 1024 * 1024

// JSONRPCVersion is the JSON-RPC protocol version used by the subprocess runtime.
const JSONRPCVersion = "2.0"

// JSONRPCRequestEnvelope is the public request metadata visible to handlers.
type JSONRPCRequestEnvelope struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// TransportHandler handles one inbound JSON-RPC request.
type TransportHandler func(context.Context, json.RawMessage, JSONRPCRequestEnvelope) (any, error)

// Transport is the bidirectional JSON-RPC runtime used by Extension.
type Transport interface {
	Handle(method string, handler TransportHandler)
	Call(ctx context.Context, method string, params any, result any) error
	Run(ctx context.Context) error
	Close() error
}

type pendingCall struct {
	result chan json.RawMessage
	err    chan error
}

type requestFrame struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  any             `json:"params,omitempty"`
}

type responseFrame struct {
	JSONRPC string              `json:"jsonrpc"`
	ID      json.RawMessage     `json:"id"`
	Result  json.RawMessage     `json:"result,omitempty"`
	Error   *JSONRPCErrorObject `json:"error,omitempty"`
}

// StdioTransport is a newline-delimited JSON-RPC 2.0 transport over stdio.
type StdioTransport struct {
	input           io.Reader
	output          io.Writer
	maxMessageBytes int

	mu       sync.Mutex
	handlers map[string]TransportHandler
	pending  map[string]pendingCall
	nextID   int64
	started  bool
	closed   bool
	err      error
	done     chan struct{}
	cancel   context.CancelFunc
	once     sync.Once
	writeMu  sync.Mutex
}

// StdioTransportOptions configures a StdioTransport.
type StdioTransportOptions struct {
	Input           io.Reader
	Output          io.Writer
	MaxMessageBytes int
}

// NewStdioTransport creates a stdio JSON-RPC transport.
func NewStdioTransport(options StdioTransportOptions) *StdioTransport {
	input := options.Input
	if input == nil {
		input = os.Stdin
	}
	output := options.Output
	if output == nil {
		output = os.Stdout
	}
	maxBytes := options.MaxMessageBytes
	if maxBytes <= 0 {
		maxBytes = DefaultMaxMessageBytes
	}
	return &StdioTransport{
		input:           input,
		output:          output,
		maxMessageBytes: maxBytes,
		handlers:        make(map[string]TransportHandler),
		pending:         make(map[string]pendingCall),
		done:            make(chan struct{}),
	}
}

// Handle registers or replaces an inbound method handler.
func (t *StdioTransport) Handle(method string, handler TransportHandler) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.handlers[strings.TrimSpace(method)] = handler
}

// Run starts the transport read loop and blocks until close or context cancellation.
func (t *StdioTransport) Run(ctx context.Context) error {
	if ctx == nil {
		return errors.New("transport: context is required")
	}
	if err := t.start(ctx); err != nil {
		return err
	}
	select {
	case <-ctx.Done():
		t.fail(ctx.Err())
		return ctx.Err()
	case <-t.done:
		t.mu.Lock()
		err := t.err
		t.mu.Unlock()
		if err == nil || errors.Is(err, io.EOF) {
			return nil
		}
		return err
	}
}

// Close closes the transport and rejects pending calls.
func (t *StdioTransport) Close() error {
	t.fail(errors.New("transport closed"))
	return nil
}

// Call sends one JSON-RPC request and decodes its response into result.
func (t *StdioTransport) Call(ctx context.Context, method string, params any, result any) error {
	if ctx == nil {
		return errors.New("transport: context is required")
	}
	cleanMethod := strings.TrimSpace(method)
	if cleanMethod == "" {
		return NewInvalidParamsError("method is required", nil)
	}
	if err := t.start(context.Background()); err != nil {
		return err
	}

	t.mu.Lock()
	if t.closed {
		err := t.err
		t.mu.Unlock()
		if err == nil {
			err = errors.New("transport closed")
		}
		return err
	}
	t.nextID++
	idBytes, err := json.Marshal(t.nextID)
	if err != nil {
		t.mu.Unlock()
		return wrapTransportError("transport: marshal request id", err)
	}
	id := json.RawMessage(idBytes)
	key := pendingKey(id)
	pending := pendingCall{result: make(chan json.RawMessage, 1), err: make(chan error, 1)}
	t.pending[key] = pending
	t.mu.Unlock()

	frame := requestFrame{
		JSONRPC: JSONRPCVersion,
		ID:      id,
		Method:  cleanMethod,
		Params:  params,
	}
	if err := t.writeFrame(frame); err != nil {
		t.removePending(key)
		return err
	}

	select {
	case raw := <-pending.result:
		if result == nil || len(raw) == 0 {
			return nil
		}
		if err := json.Unmarshal(raw, result); err != nil {
			return wrapTransportError("transport: decode response", err)
		}
		return nil
	case err := <-pending.err:
		return err
	case <-ctx.Done():
		t.removePending(key)
		return ctx.Err()
	case <-t.done:
		t.removePending(key)
		t.mu.Lock()
		err := t.err
		t.mu.Unlock()
		if err == nil {
			err = errors.New("transport closed")
		}
		return err
	}
}

func (t *StdioTransport) start(ctx context.Context) error {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.closed {
		if t.err != nil {
			return t.err
		}
		return errors.New("transport closed")
	}
	if t.started {
		return nil
	}
	if ctx == nil {
		return errors.New("transport: context is required")
	}
	loopCtx, cancel := context.WithCancel(ctx)
	started := false
	defer func() {
		if !started {
			cancel()
		}
	}()
	t.started = true
	t.cancel = cancel
	go t.readLoop(loopCtx)
	started = true
	return nil
}

func (t *StdioTransport) readLoop(ctx context.Context) {
	reader := bufio.NewReaderSize(t.input, 64*1024)
	for {
		select {
		case <-ctx.Done():
			t.fail(ctx.Err())
			return
		default:
		}
		line, err := reader.ReadBytes('\n')
		if err != nil {
			if len(line) > 0 {
				t.processLine(ctx, line)
			}
			t.fail(err)
			return
		}
		if len(line) > t.maxMessageBytes+1 {
			t.fail(fmt.Errorf("message exceeds %d bytes", t.maxMessageBytes))
			return
		}
		t.processLine(ctx, line)
	}
}

func (t *StdioTransport) processLine(ctx context.Context, line []byte) {
	trimmed := bytes.TrimSpace(line)
	if len(trimmed) == 0 {
		return
	}
	if len(trimmed) > t.maxMessageBytes {
		t.fail(fmt.Errorf("message exceeds %d bytes", t.maxMessageBytes))
		return
	}

	var envelope struct {
		JSONRPC string              `json:"jsonrpc"`
		ID      json.RawMessage     `json:"id,omitempty"`
		Method  string              `json:"method,omitempty"`
		Params  json.RawMessage     `json:"params,omitempty"`
		Result  json.RawMessage     `json:"result,omitempty"`
		Error   *JSONRPCErrorObject `json:"error,omitempty"`
	}
	if err := json.Unmarshal(trimmed, &envelope); err != nil {
		t.fail(NewRPCError(-32700, "Parse error", map[string]any{transportErrorKey: err.Error()}))
		return
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(trimmed, &fields); err != nil {
		t.fail(NewRPCError(-32700, "Parse error", map[string]any{transportErrorKey: err.Error()}))
		return
	}
	if envelope.JSONRPC != JSONRPCVersion {
		t.fail(NewInvalidRequestError("jsonrpc must be 2.0"))
		return
	}
	if envelope.Method != "" {
		if len(envelope.ID) == 0 || bytes.Equal(envelope.ID, []byte("null")) {
			return
		}
		request := JSONRPCRequestEnvelope{
			JSONRPC: envelope.JSONRPC,
			ID:      cloneRawMessage(envelope.ID),
			Method:  envelope.Method,
			Params:  cloneRawMessage(envelope.Params),
		}
		go t.dispatchRequest(ctx, request)
		return
	}
	if len(envelope.ID) > 0 {
		_, hasResult := fields["result"]
		_, hasError := fields[transportErrorKey]
		t.dispatchResponse(envelope.ID, envelope.Result, envelope.Error, hasResult, hasError)
		return
	}
	t.fail(NewInvalidRequestError("invalid json-rpc envelope"))
}

func (t *StdioTransport) dispatchRequest(ctx context.Context, request JSONRPCRequestEnvelope) {
	t.mu.Lock()
	handler := t.handlers[strings.TrimSpace(request.Method)]
	t.mu.Unlock()
	if handler == nil {
		t.sendError(request.ID, NewMethodNotFoundError(request.Method))
		return
	}
	result, err := handler(ctx, cloneRawMessage(request.Params), request)
	if err != nil {
		t.sendError(request.ID, ensureRPCError(err))
		return
	}
	t.sendResult(request.ID, result)
}

func (t *StdioTransport) dispatchResponse(
	id json.RawMessage,
	result json.RawMessage,
	rpcErr *JSONRPCErrorObject,
	hasResult bool,
	hasError bool,
) {
	key := pendingKey(id)
	t.mu.Lock()
	pending, ok := t.pending[key]
	if ok {
		delete(t.pending, key)
	}
	t.mu.Unlock()
	if !ok {
		return
	}
	switch {
	case rpcErr != nil && hasResult:
		pending.err <- NewInvalidRequestError("response must not include both result and error")
		return
	case rpcErr != nil:
		pending.err <- rpcErrorFromObject(*rpcErr)
		return
	case hasError:
		pending.err <- NewInvalidRequestError("response error must be an object")
		return
	case !hasResult:
		pending.err <- NewInvalidRequestError("response must include result or error")
		return
	default:
		pending.result <- cloneRawMessage(result)
	}
}

func (t *StdioTransport) sendResult(id json.RawMessage, result any) {
	raw, err := json.Marshal(result)
	if err != nil {
		t.sendError(id, NewInternalError(err.Error()))
		return
	}
	if len(raw) == 0 {
		raw = json.RawMessage("null")
	}
	if err := t.writeFrame(responseFrame{
		JSONRPC: JSONRPCVersion,
		ID:      cloneRawMessage(id),
		Result:  raw,
	}); err != nil {
		t.fail(err)
	}
}

func (t *StdioTransport) sendError(id json.RawMessage, err *RPCError) {
	obj := err.object()
	if writeErr := t.writeFrame(responseFrame{
		JSONRPC: JSONRPCVersion,
		ID:      cloneRawMessage(id),
		Error:   &obj,
	}); writeErr != nil {
		t.fail(writeErr)
	}
}

func (t *StdioTransport) writeFrame(frame any) error {
	encoded, err := json.Marshal(frame)
	if err != nil {
		return wrapTransportError("transport: marshal frame", err)
	}
	if len(encoded) > t.maxMessageBytes {
		return fmt.Errorf("message exceeds %d bytes", t.maxMessageBytes)
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	if _, err := t.output.Write(append(encoded, '\n')); err != nil {
		return wrapTransportError("transport: write frame", err)
	}
	return nil
}

func (t *StdioTransport) fail(err error) {
	t.once.Do(func() {
		t.mu.Lock()
		t.closed = true
		t.err = err
		cancel := t.cancel
		t.cancel = nil
		pending := t.pending
		t.pending = make(map[string]pendingCall)
		close(t.done)
		t.mu.Unlock()
		if cancel != nil {
			cancel()
		}
		for _, call := range pending {
			call.err <- err
		}
	})
}

func (t *StdioTransport) removePending(key string) {
	t.mu.Lock()
	delete(t.pending, key)
	t.mu.Unlock()
}

func pendingKey(id json.RawMessage) string {
	return string(bytes.TrimSpace(id))
}

File: sdk/typescript/src/transport.ts
import { Buffer } from "node:buffer";
import type { Readable, Writable } from "node:stream";

import {
  InvalidRequestError,
  MethodNotFoundError,
  NotInitializedError,
  ParseError,
  ensureRPCError,
  errorFromObject,
} from "./errors.js";
import type { JSONRPCID, JSONRPCRequestEnvelope, JSONRPCResponseEnvelope } from "./types.js";

export const DEFAULT_MAX_MESSAGE_BYTES = 10 * 1024 * 1024;
export const JSON_RPC_VERSION = "2.0";
export const JSON_RPC_CANCEL_METHOD = "$/cancelRequest";

export type TransportHandler = (
  params: unknown,
  request: JSONRPCRequestEnvelope,
  signal: AbortSignal
) => Promise<unknown> | unknown;

export interface TransportLike {
  call<TResult = unknown>(method: string, params?: unknown, signal?: AbortSignal): Promise<TResult>;
  handle(method: string, handler: TransportHandler): void;
  unhandle(method: string): void;
  onTransportError(listener: (error: Error) => void): () => void;
  start(): void;
  close(): Promise<void>;
}

interface PendingCall {
  resolve: (value: any) => void;
  reject: (reason: unknown) => void;
}

interface StdioTransportOptions {
  input?: Readable;
  output?: Writable;
  maxMessageBytes?: number;
}

interface RequestFrame extends JSONRPCRequestEnvelope {}

interface ResponseFrame extends JSONRPCResponseEnvelope {}

function keyOfID(id: JSONRPCID): string {
  return `${typeof id}:${String(id)}`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isRequestFrame(value: unknown): value is RequestFrame {
  if (!isRecord(value)) {
    return false;
  }
  return value.jsonrpc === JSON_RPC_VERSION && typeof value.method === "string";
}

function isResponseFrame(value: unknown): value is ResponseFrame {
  if (!isRecord(value)) {
    return false;
  }
  return value.jsonrpc === JSON_RPC_VERSION && "id" in value && !("method" in value);
}

export class StdioTransport implements TransportLike {
  private readonly input: Readable;
  private readonly output: Writable;
  private readonly maxMessageBytes: number;
  private readonly handlers = new Map<string, TransportHandler>();
  private readonly pending = new Map<string, PendingCall>();
  private readonly inbound = new Map<string, AbortController>();
  private readonly errorListeners = new Set<(error: Error) => void>();
  private readBuffer = Buffer.alloc(0);
  private nextID = 0;
  private started = false;
  private closed = false;
  private readonly onDataBound: (chunk: Buffer | string) => void;
  private readonly onErrorBound: (error: Error) => void;
  private readonly onEndBound: () => void;

  public constructor(options: StdioTransportOptions = {}) {
    this.input = options.input ?? process.stdin;
    this.output = options.output ?? process.stdout;
    this.maxMessageBytes = options.maxMessageBytes ?? DEFAULT_MAX_MESSAGE_BYTES;
    this.onDataBound = chunk => {
      this.handleChunk(typeof chunk === "string" ? Buffer.from(chunk) : chunk);
    };
    this.onErrorBound = error => {
      this.fail(error);
    };
    this.onEndBound = () => {
      this.fail(new Error("transport closed"));
    };
  }

  public start(): void {
    if (this.started) {
      return;
    }
    this.started = true;
    this.input.on("data", this.onDataBound);
    this.input.on("error", this.onErrorBound);
    this.input.on("end", this.onEndBound);
    this.input.on("close", this.onEndBound);
  }

  public handle(method: string, handler: TransportHandler): void {
    this.handlers.set(method.trim(), handler);
  }

  public unhandle(method: string): void {
    this.handlers.delete(method.trim());
  }

  public onTransportError(listener: (error: Error) => void): () => void {
    this.errorListeners.add(listener);
    return () => {
      this.errorListeners.delete(listener);
    };
  }

  public async close(): Promise<void> {
    if (this.closed) {
      return;
    }
    this.fail(new Error("transport closed"));
  }

  public async call<TResult = unknown>(
    method: string,
    params?: unknown,
    signal?: AbortSignal
  ): Promise<TResult> {
    if (this.closed) {
      throw new Error("transport closed");
    }
    this.start();

    const id = ++this.nextID;
    const frame: RequestFrame = {
      jsonrpc: JSON_RPC_VERSION,
      id,
      method,
      params,
    };

    return await new Promise<TResult>((resolve, reject) => {
      const key = keyOfID(id);
      const onAbort = (): void => {
        if (!this.pending.delete(key)) {
          return;
        }
        try {
          this.writeFrame({
            jsonrpc: JSON_RPC_VERSION,
            method: JSON_RPC_CANCEL_METHOD,
            params: { id },
          });
        } catch (error) {
          reject(error);
          return;
        }
        reject(signal?.reason ?? new DOMException("The operation was aborted", "AbortError"));
      };
      this.pending.set(key, {
        resolve: value => {
          signal?.removeEventListener("abort", onAbort);
          resolve(value);
        },
        reject: reason => {
          signal?.removeEventListener("abort", onAbort);
          reject(reason);
        },
      });
      if (signal?.aborted) {
        onAbort();
        return;
      }
      signal?.addEventListener("abort", onAbort, { once: true });
      try {
        this.writeFrame(frame);
      } catch (error) {
        this.pending.delete(key);
        signal?.removeEventListener("abort", onAbort);
        reject(error);
      }
    });
  }

  private handleChunk(chunk: Buffer): void {
    if (this.closed) {
      return;
    }

    this.readBuffer = Buffer.concat([this.readBuffer, chunk]);
    if (this.readBuffer.length > this.maxMessageBytes + 1) {
      this.fail(new Error(`message exceeds ${this.maxMessageBytes} bytes`));
      return;
    }

    while (true) {
      const newlineIndex = this.readBuffer.indexOf(0x0a);
      if (newlineIndex === -1) {
        if (this.readBuffer.length > this.maxMessageBytes) {
          this.fail(new Error(`message exceeds ${this.maxMessageBytes} bytes`));
        }
        return;
      }

      const line = this.readBuffer.subarray(0, newlineIndex);
      this.readBuffer = this.readBuffer.subarray(newlineIndex + 1);

      const trimmed = line.toString("utf8").trim();
      if (trimmed.length === 0) {
        continue;
      }
      if (Buffer.byteLength(trimmed) > this.maxMessageBytes) {
        this.fail(new Error(`message exceeds ${this.maxMessageBytes} bytes`));
        return;
      }
      this.processLine(trimmed);
      if (this.closed) {
        return;
      }
    }
  }

  private processLine(line: string): void {
    let parsed: unknown;
    try {
      parsed = JSON.parse(line);
    } catch (error) {
      this.fail(
        error instanceof Error ? new ParseError(error.message) : new ParseError(String(error))
      );
      return;
    }

    if (Array.isArray(parsed)) {
      this.fail(new InvalidRequestError("batch requests are not supported"));
      return;
    }
    if (isRequestFrame(parsed)) {
      if (parsed.method === JSON_RPC_CANCEL_METHOD) {
        this.cancelInbound(parsed.params);
        return;
      }
      if (!("id" in parsed) || parsed.id === undefined || parsed.id === null) {
        return;
      }
      this.startRequest(parsed);
      return;
    }
    if (isResponseFrame(parsed)) {
      this.dispatchResponse(parsed);
      return;
    }

    this.fail(new InvalidRequestError("invalid json-rpc envelope"));
  }

  private startRequest(request: RequestFrame): void {
    const key = keyOfID(request.id as JSONRPCID);
    this.inbound.get(key)?.abort();
    const controller = new AbortController();
    this.inbound.set(key, controller);
    void this.dispatchRequest(request, controller);
  }

  private async dispatchRequest(request: RequestFrame, controller: AbortController): Promise<void> {
    const key = keyOfID(request.id as JSONRPCID);
    const handler = this.handlers.get(request.method.trim());
    if (!handler) {
      await this.sendError(request.id as JSONRPCID, new MethodNotFoundError(request.method));
      this.inbound.delete(key);
      return;
    }

    try {
      const result = await handler(request.params, request, controller.signal);
      if (controller.signal.aborted) {
        return;
      }
      await this.sendResult(request.id as JSONRPCID, result ?? null);
    } catch (error) {
      if (controller.signal.aborted) {
        return;
      }
      await this.sendError(request.id as JSONRPCID, ensureRPCError(error));
    } finally {
      if (this.inbound.get(key) === controller) {
        this.inbound.delete(key);
      }
    }
  }

  private cancelInbound(params: unknown): void {
    if (!isRecord(params) || !("id" in params)) {
      return;
    }
    const id = params.id;
    if (typeof id !== "string" && typeof id !== "number") {
      return;
    }
    this.inbound.get(keyOfID(id))?.abort();
  }

  private dispatchResponse(response: ResponseFrame): void {
    const pending = this.pending.get(keyOfID(response.id));
    if (!pending) {
      return;
    }
    this.pending.delete(keyOfID(response.id));

    if (response.error) {
      pending.reject(errorFromObject(response.error));
      return;
    }

    pending.resolve(response.result);
  }

  private async sendResult(id: JSONRPCID, result: unknown): Promise<void> {
    this.writeFrame({
      jsonrpc: JSON_RPC_VERSION,
      id,
      result,
    } satisfies ResponseFrame);
  }

  private async sendError(id: JSONRPCID, error: Error): Promise<void> {
    const rpcError = ensureRPCError(error);
    this.writeFrame({
      jsonrpc: JSON_RPC_VERSION,
      id,
      error: rpcError.toJSONRPC(),
    } satisfies ResponseFrame);
  }

  private writeFrame(frame: RequestFrame | ResponseFrame): void {
    const encoded = JSON.stringify(frame);
    if (Buffer.byteLength(encoded) > this.maxMessageBytes) {
      throw new Error(`message exceeds ${this.maxMessageBytes} bytes`);
    }
    this.output.write(`${encoded}\n`);
  }

  private fail(error: Error): void {
    if (this.closed) {
      return;
    }
    this.closed = true;

    this.input.off("data", this.onDataBound);
    this.input.off("error", this.onErrorBound);
    this.input.off("end", this.onEndBound);
    this.input.off("close", this.onEndBound);

    for (const pending of this.pending.values()) {
      pending.reject(error);
    }
    this.pending.clear();

    for (const controller of this.inbound.values()) {
      controller.abort(error);
    }
    this.inbound.clear();

    for (const listener of this.errorListeners) {
      listener(error);
    }
  }
}

export class NotReadyTransport implements TransportLike {
  public handle(): void {}
  public unhandle(_method: string): void {}
  public onTransportError(): () => void {
    return () => {};
  }
  public start(): void {}
  public async close(): Promise<void> {}
  public async call(_method?: string, _params?: unknown): Promise<never> {
    throw new NotInitializedError();
  }
}
